File: packages/mechagent/src/mechagent/diagnostics.py
"""运行环境诊断。"""

from __future__ import annotations

import importlib.util
import shutil
import sys
from pathlib import Path
from typing import Any

from mechagent.app import MechAgent
from mechagent.core.factory import registered_meshers, registered_solvers
from mechagent.llm import LLMConfig, check_connection
from mechagent.orchestrator.capabilities import all_capabilities


def run_environment_diagnostics(
    config_path: Path = Path("config/mechagent.yaml"),
    *,
    check_llm: bool = False,
) -> dict[str, Any]:
    """检查运行 MechAgent 所需的关键环境。

    Args:
        config_path: 配置文件路径。
        check_llm: 是否调用远端 LLM 进行连接检查。

    Returns:
        dict[str, Any]: 可直接 JSON 序列化的诊断结果。
    """

    checks: list[dict[str, Any]] = [_python_check()]
    config_check, agent = _config_check(config_path)
    checks.append(config_check)
    checks.append(_package_check())
    checks.append(_studio_static_check())
    checks.append(_registry_check())
    if agent is not None:
        checks.append(_solver_check(agent))
        checks.append(_llm_check(agent, check_connection_enabled=check_llm))
    else:
        checks.append(
            _check_item(
                "solver",
                "求解器",
                False,
                True,
                {"reason": "配置未加载，无法解析求解器路径。"},
            )
        )
        checks.append(
            _check_item(
                "llm",
                "LLM 配置",
                not check_llm,
                check_llm,
                {"reason": "配置未加载，无法解析 LLM 设置。"},
            )
        )
    checks.append(_frontend_source_check())
    checks.append(_git_check())
    required = [item for item in checks if item["required"]]
    optional = [item for item in checks if not item["required"]]
    return {
        "ok": all(bool(item["ok"]) for item in required),
        "config_path": str(config_path),
        "checks": checks,
        "summary": {
            "required_passed": sum(1 for item in required if item["ok"]),
            "required_total": len(required),
            "optional_passed": sum(1 for item in optional if item["ok"]),
            "optional_total": len(optional),
        },
    }


def _python_check() -> dict[str, Any]:
    version_ok = sys.version_info >= (3, 9)
    return _check_item(
        "python",
        "Python 运行时",
        version_ok,
        True,
        {
            "executable": sys.executable,
            "version": sys.version.split()[0],
            "minimum": "3.9",
        },
    )


def _config_check(config_path: Path) -> tuple[dict[str, Any], MechAgent | None]:
    try:
        agent = MechAgent.from_config(config_path)
    except Exception as exc:
        return (
            _check_item(
                "config",
                "配置文件",
                False,
                True,
                {"path": str(config_path), "error": str(exc)},
            ),
            None,
        )
    config = agent.config
    return (
        _check_item(
            "config",
            "配置文件",
            True,
            True,
            {
                "path": str(config_path),
                "solver": config.solver.default,
                "mesher": config.mesher.default,
                "orchestrator": config.orchestrator.mode,
            },
        ),
        agent,
    )


def _package_check() -> dict[str, Any]:
    modules = [
        "fastapi",
        "gmsh",
        "httpx",
        "langgraph",
        "pydantic",
        "rich",
        "typer",
        "uvicorn",
        "yaml",
    ]
    status = {name: importlib.util.find_spec(name) is not None for name in modules}
    return _check_item(
        "packages",
        "Python 依赖",
        all(status.values()),
        True,
        {"modules": status},
    )


def _studio_static_check() -> dict[str, Any]:
    static_dir = Path(__file__).parent / "ui" / "static"
    assets_dir = static_dir / "assets"
    js_assets = sorted(item.name for item in assets_dir.glob("*.js")) if assets_dir.exists() else []
    css_assets = (
        sorted(item.name for item in assets_dir.glob("*.css")) if assets_dir.exists() else []
    )
    ok = (static_dir / "index.html").exists() and bool(js_assets) and bool(css_assets)
    return _check_item(
        "studio_static",
        "Studio 静态资源",
        ok,
        True,
        {
            "static_dir": str(static_dir),
            "index_html": (static_dir / "index.html").exists(),
            "js_assets": js_assets,
            "css_assets": css_assets,
        },
    )


def _registry_check() -> dict[str, Any]:
    solvers = registered_solvers()
    meshers = registered_meshers()
    capabilities = all_capabilities()
    return _check_item(
        "registry",
        "能力与工具注册",
        bool(solvers) and bool(meshers) and bool(capabilities),
        True,
        {
            "solvers": list(solvers),
            "meshers": list(meshers),
            "capabilities": [item.capability_id for item in capabilities],
        },
    )


def _solver_check(agent: MechAgent) -> dict[str, Any]:
    configured = agent.config.solver.calculix.path
    resolved = _resolve_executable(configured)
    return _check_item(
        "solver",
        "求解器",
        resolved is not None,
        True,
        {
            "name": agent.config.solver.default,
            "configured": configured,
            "resolved": resolved,
            "num_cpus": agent.config.solver.calculix.num_cpus,
            "timeout": agent.config.solver.calculix.timeout,
        },
    )


def _llm_check(agent: MechAgent, *, check_connection_enabled: bool) -> dict[str, Any]:
    settings = agent.config.llm
    configured = bool(settings.base_url and settings.api_key and settings.model)
    details: dict[str, Any] = {
        "base_url": settings.base_url,
        "model": settings.model,
        "api_key_present": bool(settings.api_key),
        "connection_checked": check_connection_enabled,
    }
    if check_connection_enabled and configured:
        health = check_connection(
            LLMConfig(
                base_url=settings.base_url,
                api_key=settings.api_key,
                model=settings.model,
                temperature=settings.temperature,
            )
        )
        details.update(
            {
                "connection_ok": health.ok,
                "status_code": health.status_code,
                "message": health.message,
            }
        )
        ok = health.ok
    else:
        ok = configured if check_connection_enabled else True
    return _check_item("llm", "LLM 配置", ok, check_connection_enabled, details)


def _frontend_source_check() -> dict[str, Any]:
    package_json = Path("apps/mechagent-studio/package.json")
    npm = shutil.which("npm")
    return _check_item(
        "frontend_source",
        "Studio 前端开发工具",
        package_json.exists() and npm is not None,
        False,
        {
            "package_json": str(package_json),
            "package_json_exists": package_json.exists(),
            "npm": npm,
        },
    )


def _git_check() -> dict[str, Any]:
    return _check_item(
        "git", "Git", shutil.which("git") is not None, False, {"git": shutil.which("git")}
    )


def _resolve_executable(value: str) -> str | None:
    candidate = Path(value)
    if candidate.is_absolute() or candidate.parent != Path("."):
        return str(candidate) if candidate.exists() else None
    return shutil.which(value)


def _check_item(
    key: str,
    label: str,
    ok: bool,
    required: bool,
    details: dict[str, Any],
) -> dict[str, Any]:
    return {
        "key": key,
        "label": label,
        "ok": bool(ok),
        "required": bool(required),
        "details": details,
    }

File: apps/mechagent-studio/src/lib/diagnostics.ts
import type {
  StudioDiagnosticCheck,
  StudioDiagnosticsResponse,
  StudioHealth
} from "../types";
import type { RuntimeStatus } from "./uiTypes";

export function runtimeStatus(
  diagnostics: StudioDiagnosticsResponse | null,
  health: StudioHealth | null
): RuntimeStatus {
  if (diagnostics) {
    const summary = diagnostics.summary;
    const requiredText = `必需 ${summary.required_passed}/${summary.required_total}`;
    const optionalText = `可选 ${summary.optional_passed}/${summary.optional_total}`;
    const requiredFailures = diagnosticFailureLabels(diagnostics, true);
    const optionalFailures = diagnosticFailureLabels(diagnostics, false);
    const failureText =
      requiredFailures.length > 0
        ? `；必需异常：${requiredFailures.join("、")}`
        : optionalFailures.length > 0
          ? `；可选异常：${optionalFailures.join("、")}`
          : "";
    const title = `运行环境：${requiredText}，${optionalText}；配置 ${diagnostics.config_path}${failureText}`;
    if (!diagnostics.ok) {
      return {
        ariaLabel: `运行环境异常，${requiredText}，${optionalText}`,
        label: "环境异常",
        title,
        tone: "bad"
      };
    }
    return {
      ariaLabel: `运行环境通过，${requiredText}，${optionalText}`,
      label: `环境 ${summary.required_passed}/${summary.required_total}`,
      title,
      tone: optionalFailures.length > 0 ? "warn" : "ok"
    };
  }
  if (health?.ok) {
    return {
      ariaLabel: "本地服务已连接，运行环境检测中",
      label: "环境检测中",
      title: health.config,
      tone: "neutral"
    };
  }
  return {
    ariaLabel: "本地服务检测中",
    label: "服务检测中",
    title: "等待后端健康检查返回。",
    tone: "neutral"
  };
}

export function diagnosticFailureLabels(
  diagnostics: StudioDiagnosticsResponse,
  required: boolean
): string[] {
  return diagnostics.checks
    .filter((item) => item.required === required && !item.ok)
    .slice(0, 4)
    .map((item) => item.label || item.key);
}

export function runtimeSummaryText(
  diagnostics: StudioDiagnosticsResponse | null,
  health: StudioHealth | null
): string {
  if (diagnostics) {
    const summary = diagnostics.summary;
    return `必需 ${summary.required_passed}/${summary.required_total} · 可选 ${summary.optional_passed}/${summary.optional_total}`;
  }
  return health?.ok ? "本地服务已连接，环境检测中" : "等待本地服务健康检查";
}

export function diagnosticCheckDescription(check: StudioDiagnosticCheck): string {
  const details = check.details;
  switch (check.key) {
    case "python":
      return joinRuntimeParts([detailText(details, "version"), detailText(details, "executable")]);
    case "config":
      return joinRuntimeParts([
        `solver=${detailText(details, "solver")}`,
        `mesher=${detailText(details, "mesher")}`,
        `orchestrator=${detailText(details, "orchestrator")}`
      ]);
    case "packages":
      return packageSummary(details);
    case "studio_static":
      return `JS ${detailArrayLength(details, "js_assets")} · CSS ${detailArrayLength(
        details,
        "css_assets"
      )}`;
    case "registry":
      return `solver ${detailArrayLength(details, "solvers")} · mesher ${detailArrayLength(
        details,
        "meshers"
      )} · capability ${detailArrayLength(details, "capabilities")}`;
    case "solver":
      return joinRuntimeParts([detailText(details, "name"), detailText(details, "resolved")]);
    case "llm":
      return joinRuntimeParts([
        detailText(details, "model"),
        detailBoolean(details, "api_key_present") ? "密钥已配置" : "密钥未配置",
        detailBoolean(details, "connection_checked") ? "已远端检查" : "未远端检查"
      ]);
    case "frontend_source":
      return joinRuntimeParts([detailText(details, "package_json"), detailText(details, "npm")]);
    case "git":
      return detailText(details, "git") || (check.ok ? "Git 可用" : "Git 不可用");
    default:
      return check.ok ? "通过" : "失败";
  }
}

function packageSummary(details: Record<string, unknown>): string {
  const modules = details.modules;
  if (!modules || typeof modules !== "object" || Array.isArray(modules)) {
    return "依赖状态已返回";
  }
  const entries = Object.entries(modules);
  const passed = entries.filter(([, ok]) => ok === true).length;
  const failed = entries
    .filter(([, ok]) => ok !== true)
    .map(([name]) => name)
    .slice(0, 3);
  return failed.length > 0
    ? `依赖 ${passed}/${entries.length} · 缺少 ${failed.join("、")}`
    : `依赖 ${passed}/${entries.length}`;
}

function detailText(details: Record<string, unknown>, key: string): string {
  const value = details[key];
  if (typeof value === "string") {
    return value;
  }
  if (typeof value === "number" || typeof value === "boolean") {
    return String(value);
  }
  return "";
}

function detailBoolean(details: Record<string, unknown>, key: string): boolean {
  return details[key] === true;
}

function detailArrayLength(details: Record<string, unknown>, key: string): number {
  const value = details[key];
  return Array.isArray(value) ? value.length : 0;
}

function joinRuntimeParts(parts: string[]): string {
  return parts.filter((part) => part.trim()).join(" · ");
}
